test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java

```diff
@@ -0,0 +1,1289 @@
+        }
+        return rows;
+    }
+
+    public ByteBuffer[][] getRowsAsByteBufferFromPartitions(List<BytesPartitionState> ps)
+    {
+        List<ByteBuffer[]> rows = new ArrayList<>();
+        for (BytesPartitionState p : ps)
+        {
+            for (BytesPartitionState.Row row : p.rows())
+                rows.add(getRowAsByteBuffer(p, row));
+        }
+        return rows.toArray(ByteBuffer[][]::new);
+    }
+
+    private ByteBuffer[] getRowAsByteBuffer(BytesPartitionState partition, @Nullable BytesPartitionState.Row row)
+    {
+        Clustering<ByteBuffer> pd = partition.key;
+        BytesPartitionState.Row staticRow = partition.staticRow();
+        ByteBuffer[] bbs = new ByteBuffer[factory.selectOrder.size()];
+        for (Symbol col : factory.pkPositions)
+            bbs[factory.selectOrder.indexOf(col)] = pd.bufferAt(factory.pkPositions.indexOf(col));
+        for (Symbol col : factory.staticPositions)
+            bbs[factory.selectOrder.indexOf(col)] = staticRow.get(col);
+        if (row != null)
+        {
+            for (Symbol col : factory.ckPositions)
+                bbs[factory.selectOrder.indexOf(col)] = row.clustering.bufferAt(factory.ckPositions.indexOf(col));
+            for (Symbol col : factory.regularPositions)
+                bbs[factory.selectOrder.indexOf(col)] = row.get(col);
+        }
+        return bbs;
+    }
+
+    private LookupContext context(Select select)
+    {
+        if (select.where.isEmpty())
+            throw new IllegalArgumentException("Select without a where clause is currently unsupported");
+        LookupContext ctx = new LookupContext(select);
+        context(ctx, select.where.get());
+        maybeNormalizeTokenBounds(ctx);
+        return ctx;
+    }
+
+    private void maybeNormalizeTokenBounds(LookupContext ctx)
+    {
+        if (ctx.tokenLowerBound != null && ctx.tokenUpperBound != null)
+        {
+            int rc = ctx.tokenLowerBound.token.compareTo(ctx.tokenUpperBound.token);
+            if (rc > 0)
+            {
+                // where token > 10 and < 0.... nothing matches that!
+                ctx.unmatchable = true;
+                ctx.tokenLowerBound = null;
+                ctx.tokenUpperBound = null;
+            }
+            else if (rc == 0)
+            {
+                // tokens match... but is _EQ allowed for both cases?
+                if (!(ctx.tokenLowerBound.inequality == Conditional.Where.Inequality.GREATER_THAN_EQ
+                      && ctx.tokenUpperBound.inequality == Conditional.Where.Inequality.LESS_THAN_EQ))
+                {
+                    // token < 42 and >= 42... nothing matches that!
+                    ctx.unmatchable = true;
+                    ctx.tokenLowerBound = null;
+                    ctx.tokenUpperBound = null;
+                }
+            }
+        }
+    }
+
+    private void context(LookupContext ctx, Conditional conditional)
```

Comment:
Should this be a member of `LookupContext` and maybe named something like `addConditional`?

Reply:
done